qcodes/utils/dataset/doNd.py

```diff
@@ -36,34 +37,59 @@ class UnsafeThreadingException(Exception):
     pass
 
 
-def _check_threadsafe(param_meas: Sequence[ParamMeasT]) -> None:
+class _ParamCaller:
 
-    real_parameters = [param for param in param_meas if isinstance(param, _BaseParameter)]
+    def __init__(self, *parameters: _BaseParameter):
 
-    insts = [param.root_instrument for param in real_parameters if param.root_instrument]
-    if len(set(insts)) != len(insts):
-        duplicates = [param for param in real_parameters
-                      if param.root_instrument and insts.count(param.root_instrument) > 1]
-        raise UnsafeThreadingException('Can not use threading to '
-                                       'read '
-                                       'several things from the same '
-                                       'instrument. Specifically, you '
-                                       'asked for'
-                                       ' {}.'.format(duplicates))
+        self._parameters = parameters
+
+    def __call__(self) -> Tuple[ParamDataType, ...]:
+        output = []
+        for param in self._parameters:
+            output.append(param.get())
+            output.append(param.get())
+        return tuple(output)
+
+    def __repr__(self) -> str:
+        names = tuple(param.full_name for param in self._parameters)
+        return f"ParamCaller of {names}"
+
+
+def _instrument_to_param(
+        params: Sequence[ParamMeasT]
+) -> Dict[Optional[str], Tuple[_BaseParameter, ...]]:
+
+    real_parameters = [param for param in params if isinstance(param, _BaseParameter)]
+
+    output: Dict[Optional[str], Tuple[_BaseParameter, ...]] = defaultdict(tuple)
+    for param in real_parameters:
+        if param.root_instrument:
+            output[param.root_instrument.name] += (param,)
```

Comment:
```suggestion
            output[param.root_instrument.full_name] += (param,)
```

Reply:
Sure but since it is a root instrument name == full_name

Reply:
of course, it's just that we recommend using full_name and short_name explicitly instead of name for consistency reasons. it's also fine as is.